<a href="https://colab.research.google.com/github/anurag1210/llm-zoomcamp-2025/blob/main/HomeWork2_VectorSearchforLLMZoomcamp2025.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [33]:
!pip install TextEmbedding
!pip install fastembed

In [5]:
#Importing the fastembed packages
from fastembed import TextEmbedding
import numpy as np

In [6]:
model=TextEmbedding(model_name='jinaai/jina-embeddings-v2-small-en')
query='I just discovered the course. Can I join now?'

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


Fetching 5 files:   0%|          | 0/5 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/367 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

onnx/model.onnx:   0%|          | 0.00/130M [00:00<?, ?B/s]

config.json: 0.00B [00:00, ?B/s]

In [7]:
embedding = list(model.embed([query]))[0]
print(embedding)

[-7.63946260e-02 -7.30555505e-02  5.86501763e-02  3.92670571e-02
 -1.41797115e-02 -4.68342854e-02  2.69862822e-02  3.49039060e-02
  1.53419217e-03 -1.10407363e-02  5.57556285e-02 -2.88360142e-02
 -4.87239498e-02 -9.90923923e-02  6.02579764e-02 -1.86105379e-02
 -8.81040139e-03 -3.01994831e-02 -1.42541815e-02 -3.87884212e-02
 -2.76125662e-02 -6.94957533e-03  3.20653319e-02 -5.17838394e-03
  8.37959894e-02 -8.87329606e-02 -7.30260625e-02  5.92846802e-02
  4.07805674e-02  7.72684822e-02 -4.60029697e-02  3.98426895e-02
  1.74821510e-02  8.70263712e-03 -3.09291054e-02  2.20470391e-02
  4.76479896e-02  1.76566196e-02 -3.62013627e-02 -3.53408062e-02
 -5.92537995e-03  2.97091202e-02  8.05883752e-02  1.07132722e-02
 -5.38192519e-02 -9.62881111e-03 -1.17263739e-01  3.36546374e-02
  7.05763168e-03  7.70159791e-03 -5.17482529e-02  5.46955574e-02
 -7.80289904e-02  4.21801655e-02  1.26250977e-02  5.31339061e-02
 -2.67777188e-02  1.68820697e-02 -3.75064593e-02  9.13196419e-03
  2.13696828e-02  1.60150

Q1. Embedding the query

In [10]:
print("Length of embedding:", len(embedding))

Length of embedding: 512


In [11]:
min_value_embeddings= min(embedding)
print(min_value_embeddings)

-0.11726373885183883


Cosine Similarity

Q2. Cosine similarity with another vector

In [12]:
model=TextEmbedding(model_name='jinaai/jina-embeddings-v2-small-en')
query='Can I still join the course after the start date?'

In [13]:
embedding2 = list(model.embed([query]))[0]
print(embedding2)

[-5.45303910e-02 -7.83451754e-02  3.13610346e-02  2.34234938e-02
 -3.06321480e-02 -5.51041640e-02  2.24726486e-02  1.50386159e-02
 -1.81159589e-02  1.80595354e-02 -4.80892277e-03 -7.57395955e-02
 -4.87272162e-03 -9.07717788e-02  4.83940879e-02 -4.26317430e-02
 -2.61462523e-02 -4.48129135e-02 -2.06926015e-02 -3.79054485e-02
 -3.30329342e-02 -8.02850120e-03  1.18560351e-02 -3.83721199e-02
  7.26378149e-02 -5.82811752e-02 -7.30890167e-02  3.40911381e-02
  5.60705346e-02  5.02684759e-02 -5.99293341e-02  5.33016566e-02
 -1.79524520e-02  4.17745107e-03 -2.45427902e-02  1.32288255e-02
  4.64178315e-02  1.09963765e-02 -1.01499639e-02 -5.77524144e-03
 -2.83987757e-04  4.91496354e-02  1.11214836e-01 -2.76306424e-02
 -6.78923985e-02 -1.85928261e-02 -1.23963205e-01  1.25745922e-02
  1.65298714e-02  7.22242999e-03 -2.77237445e-02  5.18073540e-02
 -5.98604791e-02  5.27339390e-02  1.03844069e-02  4.82091325e-02
 -2.29023271e-02  1.58799972e-02 -4.80511574e-02  1.02489420e-02
  1.37884538e-02  3.87444

In [14]:
cosine_similarity=np.dot(embedding,embedding2)
print(cosine_similarity)

0.9008528895674548


Q3.Ranking by Cosine
For Q3 and Q4 we will use these documents:

documents = [{'text': "Yes, even if you don't register, you're still eligible to submit the homeworks.\nBe aware, however, that there will be deadlines for turning in the final projects. So don't leave everything for the last minute.",
  'section': 'General course-related questions',
  'question': 'Course - Can I still join the course after the start date?',
  'course': 'data-engineering-zoomcamp'},
 {'text': 'Yes, we will keep all the materials after the course finishes, so you can follow the course at your own pace after it finishes.\nYou can also continue looking at the homeworks and continue preparing for the next cohort. I guess you can also start working on your final capstone project.',
  'section': 'General course-related questions',
  'question': 'Course - Can I follow the course after it finishes?',
  'course': 'data-engineering-zoomcamp'},
 {'text': "The purpose of this document is to capture frequently asked technical questions\nThe exact day and hour of the course will be 15th Jan 2024 at 17h00. The course will start with the first  “Office Hours'' live.1\nSubscribe to course public Google Calendar (it works from Desktop only).\nRegister before the course starts using this link.\nJoin the course Telegram channel with announcements.\nDon’t forget to register in DataTalks.Club's Slack and join the channel.",
  'section': 'General course-related questions',
  'question': 'Course - When will the course start?',
  'course': 'data-engineering-zoomcamp'},
 {'text': 'You can start by installing and setting up all the dependencies and requirements:\nGoogle cloud account\nGoogle Cloud SDK\nPython 3 (installed with Anaconda)\nTerraform\nGit\nLook over the prerequisites and syllabus to see if you are comfortable with these subjects.',
  'section': 'General course-related questions',
  'question': 'Course - What can I do before the course starts?',
  'course': 'data-engineering-zoomcamp'},
 {'text': 'Star the repo! Share it with friends if you find it useful ❣️\nCreate a PR if you see you can improve the text or the structure of the repository.',
  'section': 'General course-related questions',
  'question': 'How can we contribute to the course?',
  'course': 'data-engineering-zoomcamp'}]

In [15]:
  documents = [{'text': "Yes, even if you don't register, you're still eligible to submit the homeworks.\nBe aware, however, that there will be deadlines for turning in the final projects. So don't leave everything for the last minute.",
  'section': 'General course-related questions',
  'question': 'Course - Can I still join the course after the start date?',
  'course': 'data-engineering-zoomcamp'},
 {'text': 'Yes, we will keep all the materials after the course finishes, so you can follow the course at your own pace after it finishes.\nYou can also continue looking at the homeworks and continue preparing for the next cohort. I guess you can also start working on your final capstone project.',
  'section': 'General course-related questions',
  'question': 'Course - Can I follow the course after it finishes?',
  'course': 'data-engineering-zoomcamp'},
 {'text': "The purpose of this document is to capture frequently asked technical questions\nThe exact day and hour of the course will be 15th Jan 2024 at 17h00. The course will start with the first  “Office Hours'' live.1\nSubscribe to course public Google Calendar (it works from Desktop only).\nRegister before the course starts using this link.\nJoin the course Telegram channel with announcements.\nDon’t forget to register in DataTalks.Club's Slack and join the channel.",
  'section': 'General course-related questions',
  'question': 'Course - When will the course start?',
  'course': 'data-engineering-zoomcamp'},
 {'text': 'You can start by installing and setting up all the dependencies and requirements:\nGoogle cloud account\nGoogle Cloud SDK\nPython 3 (installed with Anaconda)\nTerraform\nGit\nLook over the prerequisites and syllabus to see if you are comfortable with these subjects.',
  'section': 'General course-related questions',
  'question': 'Course - What can I do before the course starts?',
  'course': 'data-engineering-zoomcamp'},
 {'text': 'Star the repo! Share it with friends if you find it useful ❣️\nCreate a PR if you see you can improve the text or the structure of the repository.',
  'section': 'General course-related questions',
  'question': 'How can we contribute to the course?',
  'course': 'data-engineering-zoomcamp'}]


In [16]:
# Step 3: Query to embed
query = "I just discovered the course. Can I join now?"

# Step 4: Generate embeddings
query_emb = list(model.embed([query]))[0]
doc_embeddings = list(model.embed([doc["text"] for doc in documents]))


In [17]:
#compute cosine similarities

cos_similarities = [np.dot(query_emb, doc_emb) for doc_emb in doc_embeddings]

In [18]:
# Step 6: Find index of highest similarity
most_similar_idx = np.argmax(cos_similarities)
print("Index with highest similarity:", most_similar_idx)

Index with highest similarity: 1


 Q4. Ranking by cosine, version two
Now let's calculate a new field, which is a concatenation of question and text:

In [19]:
full_texts = [doc['question'] + ' ' + doc['text'] for doc in documents]

In [20]:
print(full_texts)

["Course - Can I still join the course after the start date? Yes, even if you don't register, you're still eligible to submit the homeworks.\nBe aware, however, that there will be deadlines for turning in the final projects. So don't leave everything for the last minute.", 'Course - Can I follow the course after it finishes? Yes, we will keep all the materials after the course finishes, so you can follow the course at your own pace after it finishes.\nYou can also continue looking at the homeworks and continue preparing for the next cohort. I guess you can also start working on your final capstone project.', "Course - When will the course start? The purpose of this document is to capture frequently asked technical questions\nThe exact day and hour of the course will be 15th Jan 2024 at 17h00. The course will start with the first  “Office Hours'' live.1\nSubscribe to course public Google Calendar (it works from Desktop only).\nRegister before the course starts using this link.\nJoin the

In [21]:
doc_embeddings = list(model.embed(full_texts))

In [22]:
query = "I just discovered the course. Can I join now?"

In [23]:
query_embedding = list(model.embed([query]))[0]
print(len(query_embedding))

512


In [24]:
#print(doc_embeddings)
print(len(doc_embeddings))

5


In [25]:
cosine_similarities = [np.dot(query_embedding, doc_emb) for doc_emb in doc_embeddings]
print(cosine_similarities)

[np.float64(0.8514543236908068), np.float64(0.8436594159113068), np.float64(0.8408287048502558), np.float64(0.7755157969663908), np.float64(0.8086007795043938)]


In [26]:
most_similar_index = np.argmax(cosine_similarities)
max_index = np.argmax(cosine_similarities)

In [27]:
print(most_similar_index)
print(max_index)

0
0


The Value in Q3 is 1 and Q4 is 0 cause When we  include the question field in the embedding (as in Q4), that semantic connection becomes stronger, making document 0 the best match.

But in Q3, you only used the text, so document 1’s explanation about course flexibility (even after it finishes) seemed more relevant — hence it scored higher there.

Q5. Selecting the embedding model
Now let's select a smaller embedding model. What's the smallest dimensionality for models in fastembed?
One of these models is BAAI/bge-small-en. Let's use it.

In [28]:
from fastembed.embedding import DefaultEmbedding

# Use the BGE small model
model = DefaultEmbedding(model_name="BAAI/bge-small-en")

# Embed a sentence
embedding = list(model.embed(["This is a test sentence"]))[0]

print(f"Length of embedding vector: {len(embedding)}")

2025-07-01 13:10:58.000 | WARNING  | fastembed.embedding:<module>:7 - DefaultEmbedding, FlagEmbedding, JinaEmbedding are deprecated.Use from fastembed import TextEmbedding instead.


Fetching 5 files:   0%|          | 0/5 [00:00<?, ?it/s]

config.json:   0%|          | 0.00/701 [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

model_optimized.onnx:   0%|          | 0.00/133M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/366 [00:00<?, ?B/s]

Length of embedding vector: 384


The answer is 384

Q6. Indexing with qdrant (2 points)
For the last question, we will use more documents.

We will select only FAQ records from our ml zoomcamp:

In [45]:
import requests

docs_url = 'https://github.com/alexeygrigorev/llm-rag-workshop/raw/main/notebooks/documents.json'
docs_response = requests.get(docs_url)
documents_raw = docs_response.json()


documents = []

for course in documents_raw:
    course_name = course['course']
    if course_name != 'machine-learning-zoomcamp':
        continue

    for doc in course['documents']:
        doc['course'] = course_name
        documents.append(doc)

In [39]:
from fastembed import TextEmbedding
model = TextEmbedding(model_name="BAAI/bge-small-en")

In [42]:
from uuid import uuid4

texts = []
ids = []

for doc in documents:
    full_text = doc['question'] + ' ' + doc['text']
    texts.append(full_text)
    ids.append(str(uuid4()))  # Unique ID for each document

In [43]:
doc_embeddings = list(model.embed(texts))

In [44]:
print(doc_embeddings)

[array([-3.46470773e-02, -3.42522375e-02,  4.47419938e-03, -8.89333151e-03,
        2.78416295e-02,  1.58471111e-02,  1.21126715e-02,  1.04847392e-02,
       -2.27283649e-02, -5.08030839e-02,  2.74700169e-02, -2.94171125e-02,
        3.44418325e-02,  2.88361916e-03,  6.39874414e-02,  3.11142951e-02,
       -2.83699054e-02, -1.37537764e-02, -3.77648734e-02,  3.23816091e-02,
       -2.27073152e-02, -1.40944244e-02, -3.94275188e-02, -3.98991704e-02,
        2.15989370e-02,  3.46619682e-03, -1.72195025e-02, -8.53268653e-02,
       -2.16876864e-02, -1.85438722e-01, -2.63775908e-03,  1.14030403e-03,
        2.35337615e-02,  3.23457755e-02,  4.46263980e-03,  1.18379714e-02,
        1.64368581e-02, -6.32270938e-03, -3.40007781e-03,  3.60229574e-02,
       -6.63392479e-03, -9.18592140e-03, -2.97039766e-02, -5.14972508e-02,
        2.98708118e-02, -5.71320998e-03,  1.56007300e-03, -4.85710725e-02,
        1.11261467e-02,  7.28989439e-03,  5.03571471e-03, -4.50647771e-02,
       -2.86772959e-02, 

In [46]:
from qdrant_client import QdrantClient
from qdrant_client.http.models import VectorParams, Distance, PointStruct

# Start in-memory Qdrant client
client = QdrantClient(":memory:")

# Create a collection
client.recreate_collection(
    collection_name="ml-zoomcamp",
    vectors_config=VectorParams(size=len(doc_embeddings[0]), distance=Distance.COSINE),
)

# Prepare and upload points
points = []
for i, (doc, vector) in enumerate(zip(documents, doc_embeddings)):
    points.append(PointStruct(
        id=i,
        vector=vector,
        payload={
            "text": doc["text"],
            "question": doc["question"],
            "course": doc["course"]
        }
    ))

client.upsert(collection_name="ml-zoomcamp", points=points)

/tmp/ipython-input-46-3043281566.py:8: DeprecationWarning: `recreate_collection` method is deprecated and will be removed in the future. Use `collection_exists` to check collection existence and `create_collection` instead.
  client.recreate_collection(


UpdateResult(operation_id=0, status=<UpdateStatus.COMPLETED: 'completed'>)

In [47]:
query = "I just discovered the course. Can I join now?"

query_vector = list(model.embed([query]))[0]

# Search top 3 most similar documents
results = client.search(
    collection_name="ml-zoomcamp",
    query_vector=query_vector,
    limit=3
)

# Print results
for i, res in enumerate(results):
    print(f"Result {i+1}")
    print("Score:", res.score)
    print("Question:", res.payload['question'])
    print("Answer:", res.payload['text'])
    print("="*50)

Result 1
Score: 0.8703173398971558
Question: The course has already started. Can I still join it?
Answer: Yes, you can. You won’t be able to submit some of the homeworks, but you can still take part in the course.
In order to get a certificate, you need to submit 2 out of 3 course projects and review 3 peers’ Projects by the deadline. It means that if you join the course at the end of November and manage to work on two projects, you will still be eligible for a certificate.
Result 2
Score: 0.8691884279251099
Question: How long is the course?
Answer: Approximately 4 months, but may take more if you want to do some extra activities (an extra project, an article, etc)
Result 3
Score: 0.868331253528595
Question: I’m new to Slack and can’t find the course channel. Where is it?
Answer: Here’s how you join a in Slack: https://slack.com/help/articles/205239967-Join-a-channel
Click “All channels” at the top of your left sidebar. If you don't see this option, click “More” to find it.
Browse the 

/tmp/ipython-input-47-4278994388.py:6: DeprecationWarning: `search` method is deprecated and will be removed in the future. Use `query_points` instead.
  results = client.search(


In [48]:
!pip install nbstripout

NameError: name 'VectorSearchforLLMZoomcamp2025' is not defined